### ISSUE TITLE
Adding module name to method calls in procedural targets

### ISSUE THREAD

**Issue by B-rando1** [JacquesCarette/Drasil#3789]
In OO languages, when you call a method, it doesn't matter what module the object is from, because the compiler will automatically call the method for the class the object is an instance of.

In procedural languages such as Julia, method calls are being translated to function calls (with the object as the first parameter).  This means that suddenly the module the 'object' is from does matter - instead of calling `object.method()`, we need to call `module.method(object)`.

There are a few ways we can approach this problem:

### 1. Find module name automatically
In theory, we should be able to modify to GOOL's functionality to be able to check the module an object is from, and if it's different than the current module, add the module to qualify the function call.

Unfortunately, it is less easy than I had hoped to find the module name an object is from.  In the `objMethodCallMixedArgs'` function, which is the way of accessing method calls in GOOL, for the object parameter we are given an `SValue r`.
- `SValue r` has type `VS (r (Value r))`.
- Since it's wrapped in a `ValueState`, we _might_ be able to get the object's module name from the state, but I'm guessing that it would hold the current module name, not the object's module name.
- I looked at the `Value r` part as well, but it doesn't look like that would give an 'automatic' approach.  In all the renderers, this type is defined as `ValData`.  `ValData` is a record with a few things, including a `TypeData` field.  `TypeData` doesn't have anything with modules.  We could add a `moduleName` field to the `Object` type.  The problem is, types are often given by the user, so this might become tedious, especially when we have to teach `drasil-code` to add the module name to the types of objects.
- We could add a field to the State somewhere that maps object types to module names.  Then we could get the type of the object from its value, and look up that type in the State to get the module it's from.  I'm not quite sure how hard creating the mapping would be, but that could actually be pretty clean.

### 2. Require the user to give module name for method calls
We could add a `Maybe String` field to the aforementioned `objMethodCallMixedArgs'` function to hold the module name.  This is how the `call` function currently works with module names, so it's not a completely unfamiliar way of doing things.

The issue with this is that it requires the user to do extra work when creating GOOL code, just because our new targets need that information.  Because GOOL code is normally generated it might be fine, but it's not as clean as I would like.

Also, generating GOOL code in this format would require that `drasil-code` knows what module every object is from.  I'm not familiar enough with `drasil-code` to know if this is the case.  From what I've seen of it though, it is likely that `drasil-code` has more infrastructure in place for that sort of thing.

### 3. Remove module qualifiers in procedural languages
In most languages that I've seen, there's a choice to import a module using either qualified or unqualified names.  In Julia, we could do imports with `using` instead of `import`, and then we wouldn't have to worry about namespacing at all.

This does have the obvious downside that it loses out on the benefits of namespacing, however.

### 4. Remove namespacing from method calls only
This extends the ideas of **3**.  In Julia, `import` imports everything under a qualified name, but `using` imports only what is explicitly exported by the module.  Because of this, we could add only 'methods' to the export list of the module.  Then we could `import` the module so that we have everything under the qualified name, but then also `using` the module so that we have unqualified access to the 'methods'.

This technically loses out on some of the benefits of namespacing, but becuse of Julia's multiple-dispatch, this is less important for 'methods', which always take an instance of the module's class.

The downside of this approach is that it seems very Julia-specific, and likely would have problems with other procedural targets.

## Summary
I think if we can find a clean and easy way to do **1**, that would lead to the most idiomatic code, while requiring the fewest changes to GOOL code.  If not, **2** might be a reasonable compromise that would lead to equally idiomatic code.  **3** and **4** would likely be easier (at least than **1**), but would produce less idiomatic code that is less likely to be reusable for other targets.

**Comment by smiths** [comment]
@B-rando1 as usual you have produced a nice summary.  Others will have to comment on the advantages/disadvantages of each option because I'm not an expert on GOOL or Julia.  :smile:  However, a question did occur to me.  Are we trying to force Julia to look like an OO language, even though it is not?  Did you do the exercise of programming one or more of the case studies in Julia?  I believe we talked about this, but I cannot remember if it actually happened.  If we wrote the case study code in the "Julia" (multi-dispatch) kind of way, would it look different than the OO-inspired code we currently generate for the other languages?

**Comment by B-rando1** [comment]
@smiths that's definitely a fair question, and one I've come back to a few times.  I wouldn't say so much that I'm trying to make Julia look OO, as I'm trying to 'unwind' OO classes back into procedural libraries.  I wouldn't claim to be an expert of course, but in my experience with procedural languages (PHP and GML mainly) the libraries do look quite similar to classes, except that instead of methods they have functions that take an instance of a data structure as their first argument (see [here](https://www.php.net/manual/en/function.asort.php) for an example in PHP).  There are exceptions of course, but the vast majority of the time this is the pattern I've seen.  There might be more patterns available in Julia (I'll talk about that in a second), but at least in C and likely Matlab this pattern would be very standard.  Because of this, the idea of translating a class to a library feels quite natural (to me at least).

We definitely could decide not to support translating GOOL classes into Julia.  It is true that there have been more roadblocks than I expected, and while I still think it can be done, it's hard to say how many more of these discussions we'll need to have before then.  I guess why I think it's worth trying is because without support for classes, we would lose compatibility with the majority of GOOL code we're currently generating.  We could make changes to `drasil-code` to be able to generate 'procedural versions' of GOOL some of the functionality that currently requires classes (e.g. `Bundled` inputs).  The issue I have with that is that the resulting GOOL code would only be idiomatic for procedural languages, so we'd basically need to generate two sets of GOOL code: one for OO languages that uses classes, and one for procedural languages that uses structs.  If we're able to treat them as one and the same until the last step, we can remove that redundancy.  If the resulting 'procedural version' GOOL code was significantly more idiomatic for procedural languages I would consider that to be worthwhile, but as I discussed above I think that converting directly from GOOL classes can lead to code with a similar degree of idiomatic-ness.

I did get started making one of the examples in Julia in my first week or two here.  I didn't end up finishing the example, but it was a good way to learn the basics of Julia.  I think in hindsight I was probably following the Python example a bit too closely, and it is possible that doing an example without any reference material would have led to a more unique structure.  All that to say, it is possible that Julia's unique features may allow for different patterns in that language.  I can definitely look more into that, but I'm also worried that following perfectly idiomatic Julia code might lead us to create a lot of new features in GOOL that only work in Julia.

There might be more discussion to be had here, so I might open a new discussion post on the approach we want to take with translating GOOL to procedural languages.
